nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -733,73 +734,91 @@ impl StringPrototype {
     fn split(agent: &mut Agent, this_value: Value, args: ArgumentsList) -> JsResult<Value> {
         // 1. Let O be ? RequireObjectCoercible(this value).
         let o = require_object_coercible(agent, this_value)?;
-        // 2. Let S be ? ToString(O).
-        let s = to_string(agent, o)?;
 
-        // 3. Let isRegExp be ? IsRegExp(searchString).
-        // 4. If isRegExp is true, throw a TypeError exception.
-        // TODO
+        // 2. If separator is neither undefined nor null, then
+        let separator = args.get(0);
 
-        // 5. Let separatorStr be ? ToString(separatorString).
-        let separator_value = args.get(0);
+        if matches!(separator, Value::Undefined | Value::Null) {
+            // TODO: get_method(separator, Symbol.split)
+        }
 
-        // 6. If separator is undefined, return an array with this.
-        if let Value::Undefined = separator_value {
-            let slice: [Value; 1] = [this_value];
-            let results = Array::from_slice(agent, &slice);
+        // 3. Let S be ? ToString(O).
+        let s = to_string(agent, o)?;
 
-            return Ok(results.into_value());
-        }
+        // 4. If limit is undefined, lim is u32::MAX - 1.
+        let limit = args.get(0);
+        let lim = match limit {
+            Value::Undefined => u32::MAX,
+            _ => to_uint32(agent, limit)?,
+        };
 
-        let separator_str = to_string(agent, separator_value)?;
+        // 5. Let R be ? ToString(separator).
+        let r = to_string(agent, separator)?;
 
-        // 7. Let limit to be a number
-        let limit = to_number(agent, args.get(1))?;
+        // 6. If lim is zero, return an empty array
+        if lim == 0 {
+            let list: [Value; 0] = [];
+            return Ok(create_array_from_list(agent, &list).into_value());
+        }
 
-        // 8. If the limit is zero, returns an empty array
-        if limit.is_pos_zero(agent) || limit.is_neg_zero(agent) {
-            let a = array_create(agent, 0, 0, None)?;
-            return Ok(a.into_value());
+        // 7. If separator is undefined, return an array with the whole string
+        if let Value::Undefined = separator {
+            let list: [Value; 1] = [s.into_value()];
+            return Ok(create_array_from_list(agent, &list).into_value());
         }
 
-        // 9. Split the string
-        let subject = s.as_str(agent).to_owned();
-        let separator = separator_str.as_str(agent).to_owned();
-        let split = subject.split(&separator);
+        // 8. Let separatorLength be the length of R.
+        let separator_length = r.len(agent);
 
-        // 10. The limit must be positive integer or 0
-        let limit = if limit.is_sign_positive(agent) {
-            limit.into_i64(agent) as usize
-        } else {
-            0
-        };
+        // 9. If separatorLength = 0, the split by characters
+        if separator_length == 0 {
+            let subject = s.as_str(agent).to_owned();
```

Comment:
**nitpick:** I would leave in the spec steps a through e, BUT also **praise:** Explaining that this is doing a split by characters is very nice!

**suggestion:** I _think_ we're guaranteed that all of the characters are necessarily expressible within 7 bytes as UTF-8. I think so anyway. As such, I think you can actually avoid the `to_owned()` here by just asserting that each split part will pass `SmallString::try_from(part).unwrap()`.

There would also be a way to avoid the Vec altogether and just shuffle data within the heap directly, but we're missing a `ArrayHeapMutIndexable` struct for now. `map_constructor.rs:172` and the `ArrayHeapIndexable` struct may serve as hints as to what I'm thinking about here. Just add in the knowledge that any UTF-16 character (which is what we should be splitting the string into) should fit into 7 bytes when converted into UTF-8, I think, and all should be "clear" :D

Reply:
I've refactored this part.